File: lex/lexer.py
from typing import List
from . import FileLexer, Token, IdentInfo, Tok, Loc

class Lexer:
    cur_lexer: FileLexer | None
    lexer_stack: List[FileLexer]
    cached_tokens: List[Token]
    include_paths: List[str]

    def __init__(self):
        self.cur_lexer = None
        self.lexer_stack = []
        self.cached_tokens = []
        self.include_paths = []

    def add_include_paths(self, paths):
        self.include_paths += paths

    def enter_token(self, token: Token, a: bool):
        self.cached_tokens.append(token)

    def enter_source_file(self, filename: str):
        if self.cur_lexer is not None:
            self.lexer_stack.append(self.cur_lexer)
        self.cur_lexer = FileLexer(filename, self)

    def end_source_file(self, token: Token):
        if self.cur_lexer is None:
            return
        if len(self.lexer_stack) == 0:
            self.cur_lexer.construct_token(token, self.cur_lexer.pos + 1, Tok.EOF)
            return
        self.cur_lexer = self.lexer_stack[-1]
        self.lexer_stack.pop()
        tmp_tok = self.cur_lexer.lex()
        token.copy_from(tmp_tok)

    def find_file_to_include(self, filename: str, loc: Loc) -> str:
        from utils.diagnostic import diag, Diag
        import os.path
        if os.path.isfile(filename):
            return filename
        for ip in self.include_paths:
            f = ip + "/" + filename
            if os.path.isfile(f):
                return f
        diag(loc, f"include file not found: '{filename}'", Diag.ERROR)
        return "/dev/null"

    def handle_directive(self, token: Token):
        dir_kind = self.cur_lexer.read_to_whitespace()
        assert dir_kind == "include", f"Only include directive supported (not {dir_kind})"
        include_file = self.cur_lexer.read_to_end_of_line().strip()
        include_file = self.find_file_to_include(include_file, token.loc)
        self.lexer_stack.append(self.cur_lexer)
        self.cur_lexer = FileLexer(include_file, self)
        token.copy_from(self.lex())

    def handle_ident(self, token: Token, ident_info: IdentInfo):
        pass

    def lex(self) -> Token:
        if len(self.cached_tokens) != 0:
            token = self.cached_tokens[0]
            self.cached_tokens = self.cached_tokens[1:]
            return token
        if self.cur_lexer is None:
            assert False, "Nothing to lex"
        return self.cur_lexer.lex()

File: cppver/compiler/lib/lexer/lexer.cpp
#include "lexer.hpp"
#include "diags/diagnostic.hpp"
#include "lexer/tok.hpp"
#include <algorithm>
#include <cassert>
#include <cctype>
#include <filesystem>
#include <memory>

void Lexer::enter_source_file(std::string const &filename) {
  if (cur_lexer != nullptr) {
    lexer_stack.push_back(std::move(cur_lexer));
  }
  cur_lexer = std::make_unique<FileLexer>(filename, this);
}

void Lexer::end_source_file(Token &tok, u32 cur_pos) {
  if (cur_lexer == nullptr)
    return;
  if (lexer_stack.size() == 0) {
    cur_lexer->construct_token(tok, cur_pos, tok::END_OF_FILE);
    return;
  }
  cur_lexer = std::move(lexer_stack.back());
  lexer_stack.pop_back();
  tok = lex();
}

std::string_view include_directive_strip_filename(std::string_view filename) {
  while (std::isspace(filename.front())) {
    filename = filename.substr(1);
  }
  while (std::isspace(filename.back())) {
    filename = filename.substr(0, filename.length() - 1);
  }
  return filename;
}

std::string find_file_to_include(Lexer &l, std::string_view filename, Loc loc) {
  namespace fs = std::filesystem;
  fs::path path(filename);
  if (fs::exists(path) && !fs::is_directory(path)) {
    return std::string(filename);
  }
  if (path.is_absolute()) {
    Diag(diag::ERROR, loc, "file not found for absolute path '%.*s'",
         filename.length(), filename.data());
    return "/dev/null";
  }
  fs::path cur_filename(l.cur_lexer ? l.cur_lexer->f->file_name : "");
  if (cur_filename.has_parent_path()) {
    fs::path dir_path = cur_filename.parent_path();
    fs::path lookup_path = dir_path / path;
    if (fs::exists(lookup_path) && !fs::is_directory(lookup_path)) {
      return lookup_path.string();
    }
  }
  for (auto const &ip : l.include_paths) {
    fs::path dir_path(ip);
    fs::path lookup_path = dir_path / path;
    if (fs::exists(lookup_path) && !fs::is_directory(lookup_path)) {
      return lookup_path.string();
    }
  }
  Diag(diag::ERROR, loc, "file not found for path '%.*s'", filename.length(),
       filename.data());
  return "/dev/null";
}

void Lexer::handle_directive(Token &tok) {
  std::string dir_kind = cur_lexer->read_to_whitespace();
  if (dir_kind != "include") {
    Diag(diag::ERROR, tok.loc, "unsupported directive '%s'", dir_kind.c_str());
    cur_lexer->read_to_end_of_line();
    tok = lex();
    return;
  }
  std::string file_to_include = cur_lexer->read_to_end_of_line();
  std::string_view file_to_include_stripped =
      include_directive_strip_filename(file_to_include);
  file_to_include =
      find_file_to_include(*this, file_to_include_stripped, tok.loc);
  if (std::find(already_included_files.begin(), already_included_files.end(),
                file_to_include) == already_included_files.end()) {
    already_included_files.push_back(file_to_include);
    enter_source_file(file_to_include);
  }
  tok = lex();
}

void Lexer::handle_ident(Token &tok, IdentInfo *ident) {
  (void)tok, (void)ident;
}

Token Lexer::lex() {
  if (cached_tokens.size() != 0) {
    auto t = cached_tokens.front();
    cached_tokens.erase(cached_tokens.begin());
    return t;
  }
  assert(cur_lexer != nullptr && "Nothing to lex");
  return cur_lexer->lex();
}
